x/virtualgroup/keeper/keeper.go

```diff
@@ -650,3 +667,215 @@ func (k Keeper) CompleteSwapOut(ctx sdk.Context, gvgFamilyID uint32, gvgIDs []ui
+		GlobalVirtualGroupFamilyId: gvgFamilyID,
+		GlobalVirtualGroupId:       gvgID,
+		TargetSpId:                 swapInInfo.TargetSpId,
+	}); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (k Keeper) CompleteSwapIn(ctx sdk.Context, gvgFamilyID uint32, gvgID uint32, successorSP *sptypes.StorageProvider) error {
+	store := ctx.KVStore(k.storeKey)
+	swapInInfo := types.SwapInInfo{}
+	// swapIn family
+	if gvgFamilyID != types.NoSpecifiedFamilyId {
+		key := types.GetSwapInFamilyKey(gvgFamilyID)
+		bz := store.Get(key)
+		if bz == nil {
+			return types.ErrSwapInFailed.Wrapf("The swap info not found in blockchain.")
+		}
+		k.cdc.MustUnmarshal(bz, &swapInInfo)
+		if successorSP.Id != swapInInfo.SuccessorSpId {
+			return types.ErrSwapInFailed.Wrapf("The SP(ID: %d) has not reserved the swap(swapInfo=%s)", successorSP.Id, swapInInfo.String())
+		}
+		if uint64(ctx.BlockTime().Unix()) >= swapInInfo.ExpirationTime {
+			return types.ErrSwapInFailed.Wrapf("reserved swap expired")
+		}
+
+		targetPrimarySP, found := k.spKeeper.GetStorageProvider(ctx, swapInInfo.TargetSpId)
+		if !found {
+			return sptypes.ErrStorageProviderNotFound.Wrapf("The storage provider(ID: %d) not found when complete swap in.", swapInInfo.TargetSpId)
+		}
+		if err := k.SwapAsPrimarySP(ctx, targetPrimarySP, successorSP, gvgFamilyID, true); err != nil {
+			return err
+		}
+		store.Delete(key)
+	} else {
+		key := types.GetSwapInGVGKey(gvgID)
+		bz := store.Get(key)
+		if bz == nil {
+			return types.ErrSwapInFailed.Wrapf("The swap info not found in blockchain.")
+		}
+		k.cdc.MustUnmarshal(bz, &swapInInfo)
+		if successorSP.Id != swapInInfo.SuccessorSpId {
+			return types.ErrSwapInFailed.Wrapf("The sp(ID: %d) has not reserved the swap for secondary SP(ID: %d)", successorSP.Id, swapInInfo.TargetSpId)
+		}
+		targetSecondarySP, found := k.spKeeper.GetStorageProvider(ctx, swapInInfo.TargetSpId)
+		if !found {
+			return sptypes.ErrStorageProviderNotFound.Wrapf("The storage provider(ID: %d) not found when complete swap in.", swapInInfo.TargetSpId)
+		}
+		if err := k.completeSwapInGVG(ctx, successorSP.Id, targetSecondarySP.Id, gvgID); err != nil {
+			return err
+		}
+		store.Delete(key)
+	}
+	if err := ctx.EventManager().EmitTypedEvents(&types.EventCompleteSwapIn{
+		StorageProviderId:          successorSP.Id,
+		TargetStorageProviderId:    swapInInfo.TargetSpId,
+		GlobalVirtualGroupFamilyId: gvgFamilyID,
+		GlobalVirtualGroupId:       gvgID,
+	}); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (k Keeper) completeSwapInGVG(ctx sdk.Context, successorSPID, targetSecondarySPID uint32, gvgID uint32) error {
+	gvg, found := k.GetGVG(ctx, gvgID)
+	if !found {
+		return types.ErrGVGNotExist
+	}
+	if successorSPID == gvg.PrimarySpId {
+		return types.ErrSwapInFailed.Wrapf("the primary SP(ID: %d) can not swap into GVG's secondary (%s).", successorSPID, gvg.String())
+	}
+
+	// settlement
+	err := k.SettleAndDistributeGVG(ctx, gvg)
```

Comment:
You have done the settlement, but did not disconnect the paytment stream with old SP and build new payment stream to new sp.